### PR TITLE
[Merged by Bors] - node: check genesis accounts before creating

### LINKED ISSUE #3303: node restart fails on committing genesis state
## Description

when the node aborted before it records the latest layer (but already generated the genesis accounts), upon restarts it will try to create genesis account again and fail as below. we need to find a better signal for whether to regenerate the genesis account.

```
{"L":"FATAL","T":"2022-06-28T19:13:00.129+0300","N":"00000.defaultLogger","M":"Failed to run the node. See logs for details.","errmsg":"start node: cannot start services: setup genesis: cannot commit genesis state: failed to insert account [255 8 60 154 34 224 93 52 89 176 63 61 190 214 28 122 214 224 162 9] for layer 119: database: object exists","name":""}
```

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation

Closes #3303 and some minor fixes in mesh

## Changes

- check whether the first genesis account exists (has balance). only try to create the genesis accounts when it's missing
- simplify mesh initialization
- remove one extra SetApplied call in mesh.go

## Test Plan

unit test, systest, manual testing with node+devnet

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
> ## try
> Build failed:
> 
> * [ci-stage2](https://github.com/spacemeshos/go-spacemesh/runs/7155975071?check_suite_focus=true)

https://github.com/spacemeshos/go-spacemesh/issues/3316

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
> ## try
> Build failed:
> 
> * [ci-stage2](https://github.com/spacemeshos/go-spacemesh/runs/7156770115?check_suite_focus=true)

https://github.com/spacemeshos/go-spacemesh/issues/3316 again

**Review by evt — APPROVED** [review]

**Comment by countvonzero** [comment]
bors merge

**Comment by noamnelke** [comment]
bors try

**Review by abergasov — APPROVED** [review]

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge

File: cmd/node/node.go
Comment (jonZlotnik) [inline cmd/node/node.go]: Why do we use index 0 to check error? Possible to avoid, or abstract away magic numbers?

File: cmd/node/node.go
Comment (jonZlotnik) [inline cmd/node/node.go]: Don't really like this "try-catch" nested error logic going on. Can we make this more linear? Hard to read.

File: mesh/mesh.go
Comment (jonZlotnik) [inline mesh/mesh.go:61]: I agree with less return values, but do are we losing that "recovery checkability"? Was it useful in the first place?

File: mesh/mesh.go
Comment (jonZlotnik) [inline mesh/mesh.go:69]: yaas. clean and descriptive!

File: mesh/mesh.go
Comment (jonZlotnik) [inline mesh/mesh.go:84]: need better variable name. Also, ID should be capitalized https://github.com/golang/go/wiki/CodeReviewComments#initialisms

Long but descriptive is always better than short but mysterious.
I propose `gensisLayerID`  or `gLyrID` I don't really care, but `gLid` is just too unreadable.
Just be sure to keep the file consistent. Note: `gLayer` on line 101

Comment (jonZlotnik) [reply mesh/mesh.go:84]: Also, we should change `types.GetEffectiveGenesis()` to `types.EffectiveGenesisLayerID()`
- more descriptive,
- follows [guidelines](https://go.dev/doc/effective_go#Getters)
"it's neither idiomatic nor necessary to put Get into the getter's name"

File: mesh/mesh_test.go
Comment (jonZlotnik) [inline mesh/mesh_test.go:99]:
```suggestion
func requireLastAppliedInDB(t *testing.T, mesh *Mesh, expected types.LayerID) {
```

File: mesh/mesh_test.go
Comment (jonZlotnik) [inline mesh/mesh_test.go:109]:
```suggestion
func requireLatestInDB(t *testing.T, mesh *Mesh, expected types.LayerID) {
```

File: mesh/mesh_test.go
Comment (jonZlotnik) [inline mesh/mesh_test.go:116]:
```suggestion
func requireProcessedInDB(t *testing.T, mesh *Mesh, expected types.LayerID) {
```

File: mesh/mesh_test.go
Comment (jonZlotnik) [inline mesh/mesh_test.go:128]:
```suggestion
	latestVerified, err := tm.Mesh.GetVerifiedLayer()
```

File: mesh/mesh_test.go
Comment (jonZlotnik) [inline mesh/mesh_test.go:125]:
```suggestion
	latestProcessed, err := tm.Mesh.GetProcessedLayer()
```

File: mesh/mesh_test.go
Comment (jonZlotnik) [inline mesh/mesh_test.go:131]:
```suggestion
	latest := tm.Mesh.LatestLayer()
```

File: mesh/mesh_test.go
Comment (jonZlotnik) [inline mesh/mesh_test.go:168]: again, the "got" thing is weird

File: cmd/node/node.go
Comment (jonZlotnik) [reply cmd/node/node.go]: The code isn't self-explanatory. (I needed to read the PR desc again)

File: cmd/node/node.go
Comment (countvonzero) [reply cmd/node/node.go]: would you like to make a suggestion?

File: mesh/mesh.go
Comment (countvonzero) [reply mesh/mesh.go:61]: because it's not used anymore.

this change is about using a different way to decide whether we need to re-create genesis account.

File: mesh/mesh.go
Comment (countvonzero) [reply mesh/mesh.go:84]: there is no established guideline for variable naming as of now. and i don't really want to debate this now.
if you want, try to get ppl to agree on a guideline and i am happy to follow. same for other renaming feedback.

File: cmd/node/node.go
Comment (countvonzero) [reply cmd/node/node.go]: because the genesis accounts are created atomically in a database transaction.

Comment (jonZlotnik) [reply cmd/node/node.go]: But why do we only care about the first one?

File: cmd/node/node.go
Comment (jonZlotnik) [inline cmd/node/node.go:467]: What does it mean if len(genesisAccts) == 0 ?

File: cmd/node/node.go
Comment (countvonzero) [reply cmd/node/node.go]: because the first one is representative of rest of it. they are created atomically.
if one exists, all exist.
if one doesn't exist, none exist.

File: cmd/node/node.go
Comment (jonZlotnik) [reply cmd/node/node.go]: Nope. Just complaining.
If I think of something actually cleaner before this merges, I'll post.

File: cmd/node/node.go
Comment (countvonzero) [reply cmd/node/node.go:467]: for genesis, we should probably fail without genesis accounts. but i'll leave that to a different PR, since i'm not modifying that behavior here.

for now, some systests do not set genesis accounts.

File: cmd/node/node.go
Comment (jonZlotnik) [inline cmd/node/node.go]: Don't know how many genesis accounts are expected, but should we not use the `genesisAccts` alrady initialized on line 466 ?

**Comment by countvonzero** [comment]
@dshulyak 

as discussed in slack, added AccountExists to vm. https://github.com/spacemeshos/go-spacemesh/pull/3312/commits/cf29b4f106e22eec86158da4ad2e969e87c8bb43

**Comment by countvonzero** [comment]
bors try

**Review by dshulyak — APPROVED** [review]

**Comment by countvonzero** [comment]
bors merge

**Pull request closed without merging**
